### DISCUSSION TITLE
Graph-dependent matrix multiplication in a batch

### DISCUSSION THREAD

**Discussion by Nokimann** [pyg-team/pytorch_geometric#6767] [Q&A]
Let's assume that there are `n` graphs in a batch.

I'm trying to multiply `pos` in a batch with `n` 3x3 matrices corresponding to each graph where matrices are different from each other with different graphs.

How can we do this matrix multiplication without `repeat_interleave` function?

**Comment by Nokimann** [comment]
I found [segment_matmul](https://github.com/pyg-team/pyg-lib/blob/master/pyg_lib/csrc/ops/matmul.h) in source of [HeteroLinear](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=heterolinear#torch_geometric.nn.dense.HeteroLinear) seems to work. But, does it efficiently work?

Then, how can I load `segment_matmul`?

And does it work like the below?

```
pos.unsqueeze(1) @ matrix[batch]
```

**Reply by rusty1s** [reply]
Yes, `segment_matmul` seems to be perfect for this. It is optimized by both Intel and NVIDIA so it is as fast as it can get :)

**Reply by Nokimann** [reply]
@rusty1s 
But, How can I load `segment_matmul`?

I just followed the source code of [HeteroLinear](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=heterolinear#torch_geometric.nn.dense.HeteroLinear), but failed to load `segment_matmul` 

```
from torch_geometric.typing import pyg_lib

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[39], line 1
----> 1 from torch_geometric.typing import pyg_lib

ImportError: cannot import name 'pyg_lib' from 'torch_geometric.typing' (.../lib/python3.10/site-packages/torch_geometric/typing.py)
```

**Reply by rusty1s** [reply]
Do you have PyG 2.2 installed?

**Reply by Nokimann** [reply]
@rusty1s 

```
$ conda list | grep pyg
pyg                       2.2.0           py310_torch_1.13.0_cu116    pyg
```
Maybe, I have to reinstall my environment?

**Reply by rusty1s** [reply]
My bad, you need to look at the source code of PyG 2.2: https://pytorch-geometric.readthedocs.io/en/stable/modules/nn.html#torch_geometric.nn.dense.HeteroLinear

**Reply by Nokimann** [reply]
So, it is not implemented yet? hope to update soon...

**Reply by rusty1s** [reply]
It is integrated, the imports just changed from PyG 2.2 to master. Instead of doing `from torch_geometric.typing import pyg_lib`, you would just need to write `import pyg_lib`
